design/KEPS/namespace-discovery/proposal.md

```diff
@@ -12,19 +12,116 @@ Provide a better way for namespace discovery within both server and operator.
+## Istio Discovery Selectors
+
+Kiali already autodiscovers the istio controlplanes. Discovery selectors for the istio controlplane can be found by looking at the controlplane's `istio` configmap under `mesh.discoverySelectors`. Kiali will read this configmap and if discovery selectors have been set, Kiali will scope itself to the namespaces selected by the discovery selectors. In cases where there are multiple controlplanes e.g. with revisioned upgrades, Kiali will read both discovery selectors and OR the two together scoping itself to the disjoint of the two sets. For example:
+
+Controlplane A discovery selector:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: api
+```
+
+Controlplane B discovery selector:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: backend
+```
+
+Kiali discovery selector:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: backend
+  - matchLabels:
+      team: api
+```
+
+For multi-primary cluster deployments, Kiali will scope the namespaces according to the discovery selector's defined on that cluster.
+
+Cluster A discovery selector:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: api
+```
+
+Cluster B discovery selector:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: backend
+```
+
+Kiali discovery selector cluster A:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: api
+```
+
+Kiali discovery selector cluster B:
+
+```yaml
+discoverySelectors:
+  - matchLabels:
+      team: backend
+```
+
+The same rules for multiple revisions in a multi-cluster deployment would still apply.
+
+Primary-remote deployments would behave the same as single cluster where the remote cluster is scoped to the primary controlplane's discovery selectors.
+
+TODO: external controlplane
+
+### Failure scenarios
+
+Since the discovery selectors act as a way to separate multiple Kiali deployments from one another, when Kiali relies on autodiscovering the Istio discovery selectors and it cannot auto discover them, the Kiali server will fail to start and/or panic rather than potentially show out of scope data. In a multi-primary deployment, if one primary cannot be reached then that cluster will be considered inaccessible.
+
+## Discovery Selector Kiali Config Option
+
+A new configuration option will be added that will mimic the Istio discovery selectors configuration option. Using the configuration option, users will be able to specify which namespaces Kiali should be scoped to. Having a config option within Kiali has the advantage of not relying on autodiscovery of the Istio controlplanes. In cases where Kiali temporarily cannot access the control plane configuration, Kiali does not need to fail since it will rely on its own discovery selectors to scope itself. The configuration option needs to provide enough flexibility for users to specify different discovery selectors for each cluster in multi-primary scenarios.
```

Comment:
Question is default behavior - IF the user does not set the  Kiali DS, what is the default behavior? I assume it is, "Use the Istio DS." If Istio DS cannot be found, then do what? Abort with error, I assume.

If the user DOES specify Kiali DS, then I assume Istio DS is ignored entirely.

Reply:
Do we need an option to use Istio DS *if possible*?

Reply:
I don't think so, unless a nil Kiali DS can have more than one semantic. I assume you either set Kiali DS and ignore Istio DS, or you do NOT set Kiali DS and you use Istio DS.

When would you set Kiali DS AND STILL want to use Istio DS?
When would you NOT set Kiali DS AND NOT want to use Istio DS either?

Reply:
> Question is default behavior - IF the user does not set the Kiali DS, what is the default behavior? I assume it is, "Use the Istio DS." If Istio DS cannot be found, then do what? Abort with error, I assume.

Yes.

> If the user DOES specify Kiali DS, then I assume Istio DS is ignored entirely.

Yes.

> Do we need an option to use Istio DS if possible?

Something like "fallback to Kiali DS but try to use Istio DS first"? That would be useful but I also don't want to add more config options than we have to.

Reply:
> Something like "fallback to Kiali DS but try to use Istio DS first"? 

Oh, that's an interesting one. Er... I agree I don't want more config than we have to. Is there a real use-case for this "try Istio DS first, even if Kiali DS is set"?